### PR TITLE
Tutorial pagination

### PR DESCRIPTION
**Pull request by recursiveforte**
add pagination to tutorials + add sprig batch jams content + change tutorial file name scheme

### CODE REVIEW

**Comment by kognise** [comment]
is there somewhere i can see a tutorial that has pagination so i can preview this?

File: games/getting_started-0.md [hunk 1]
```diff
@@ -228,4 +228,6 @@ You’re now a Sprig game developer! What games do you want to build? Not sure w
 - having different boxes and goal types.
 - come up with your own mechanic!
 
+Once you're finished with this game, check out parts two, three, and four of the larger [Sprig Batch Jam](https://jams.hackclub.com/batch/sprig)
```
Comment (kognise) [inline games/getting_started-0.md:231]: period for grammar!

```suggestion
Once you're finished with this game, check out parts two, three, and four of the larger [Sprig Batch Jam](https://jams.hackclub.com/batch/sprig).
```

File: src/components/popups-etc/help.tsx [hunk 2]
```diff
@@ -1,13 +1,14 @@
 import { Signal, useSignal, useSignalEffect } from '@preact/signals'
-import { IoCaretDown, IoCaretUp } from 'react-icons/io5'
+import {IoCaretBack, IoCaretDown, IoCaretForward, IoCaretUp} from 'react-icons/io5'
```
Comment (kognise) [inline src/components/popups-etc/help.tsx:2]:
```suggestion
import { IoCaretBack, IoCaretDown, IoCaretForward, IoCaretUp } from 'react-icons/io5'
```

don't break the formatting!

File: src/components/popups-etc/help.tsx [hunk 3]
```diff
@@ -1,13 +1,14 @@
...
 import styles from './help.module.css'
 import { compiledContent } from '../../../docs/docs.md'
-import { PersistenceState } from '../../lib/state'
+import {codeMirror, PersistenceState} from '../../lib/state'
```
Comment (kognise) [inline src/components/popups-etc/help.tsx:5]:
```suggestion
import { codeMirror, PersistenceState } from '../../lib/state'
```

File: src/components/popups-etc/help.tsx [hunk 4]
```diff
@@ -1,13 +1,14 @@
...
 import Button from '../design-system/button'
+import {saveGame} from "../big-interactive-pages/editor";
```
Comment (kognise) [inline src/components/popups-etc/help.tsx:7]:
```suggestion
import { saveGame } from '../big-interactive-pages/editor'
```

File: src/components/popups-etc/help.tsx [hunk 5]
```diff
@@ -17,11 +18,44 @@ export default function Help(props: HelpProps) {
 	const visible = useSignal(props.tutorialContent ? true : (props.initialVisible ?? false))
 	const showingTutorial = useSignal(props.tutorialContent !== undefined)
 
-	const tutorialHtml = props.tutorialContent
+	const tutorialHtml = props.tutorialContent 
+		&& (props.persistenceState?.value.kind == "PERSISTED" || props.persistenceState?.value.kind == "SHARED")
```
Comment (kognise) [inline src/components/popups-etc/help.tsx:22]: code style matters!

```suggestion
		&& (props.persistenceState?.value.kind == 'PERSISTED' || props.persistenceState?.value.kind == 'SHARED')
```

File: src/components/popups-etc/help.tsx [hunk 6]
```diff
@@ -17,11 +18,44 @@ export default function Help(props: HelpProps) {
...
+			props.persistenceState.value.tutorialIndex || 0
+		setTutorialIndex(tutorialIndex + 1)
+	}
+	
+	const previousPage = () => {
+		const tutorialIndex =
+			(props.persistenceState?.value.kind == "PERSISTED" || props.persistenceState?.value.kind == "SHARED") &&
+			props.persistenceState.value.tutorialIndex || 0
+		setTutorialIndex(tutorialIndex - 1)
+	}
```
Comment (kognise) [inline src/components/popups-etc/help.tsx:58]:
```suggestion
	const setTutorialIndex = (tutorialIndex: number) => {
		if (props.persistenceState?.value.kind == 'PERSISTED') {
			props.persistenceState.value = {
				...props.persistenceState.value,
				stale: true,
				cloudSaveState: 'SAVING',
				tutorialIndex
			}
			saveGame(props.persistenceState, codeMirror.value!.state.doc.toString())
		} else if (props.persistenceState?.value.kind == 'SHARED') {
			props.persistenceState.value = {
				...props.persistenceState.value,
				tutorialIndex
			}
		}
	}
	
	const nextPage = () => {
		const tutorialIndex = 
			(props.persistenceState?.value.kind == 'PERSISTED' || props.persistenceState?.value.kind == 'SHARED') && 
			props.persistenceState.value.tutorialIndex || 0
		setTutorialIndex(tutorialIndex + 1)
	}
	
	const previousPage = () => {
		const tutorialIndex =
			(props.persistenceState?.value.kind == 'PERSISTED' || props.persistenceState?.value.kind == 'SHARED') &&
			props.persistenceState.value.tutorialIndex || 0
		setTutorialIndex(tutorialIndex - 1)
	}
```

File: src/components/popups-etc/help.tsx [hunk 7]
```diff
@@ -44,11 +78,35 @@ export default function Help(props: HelpProps) {
...
 			{tutorialHtml && props.persistenceState && visible.value && showingTutorial.value && (
 				<div class={styles.content} >
-					<div dangerouslySetInnerHTML={{ __html: tutorialHtml }} />
-					{props.persistenceState.value.kind === 'PERSISTED' && (
-						<Button onClick={() => {
-							props.showingTutorialWarning!.value = true
-						}} accent>Exit Tutorial</Button>
+					<div dangerouslySetInnerHTML ={{ __html: tutorialHtml }} />
+
+					{(props.persistenceState?.value.kind == "PERSISTED" || props.persistenceState?.value.kind == "SHARED") && (
```
Comment (kognise) [inline src/components/popups-etc/help.tsx:83]:
```suggestion
					{(props.persistenceState?.value.kind == 'PERSISTED' || props.persistenceState?.value.kind == 'SHARED') && (
```

File: games/maze_game_starter-0.md
Comment (kognise) [inline games/maze_game_starter-0.md:1]: Just noting that I am *not* reviewing any of the tutorial copy. Let me know if you want me to do so.

File: games/metadata.json [hunk 8]
```diff
@@ -2838,5 +2838,13 @@
     "img": "thunderball",
     "tags": [],
     "added on": "2023-08-03"
+  },
+  {
+    "filename": "maze_game_starter",
+    "title": "maze_game_starter",
+    "author": "Cheru Berhanu",
+    "img": "",
+    "tags": [],
```
Comment (kognise) [inline games/metadata.json:2847]: This should have tags, especially the tutorial one if it is a tutorial.

File: src/components/popups-etc/help.module.css
Comment (kognise) [inline src/components/popups-etc/help.module.css:1]: Please decrease padding around the dialog.

File: src/components/popups-etc/help.tsx [hunk 9]
```diff
@@ -44,11 +78,35 @@ export default function Help(props: HelpProps) {
...
+							&& props.persistenceState.value.tutorialIndex < props.persistenceState.value.tutorial.length - 1
+							&& (<Button onClick={nextPage} class={styles.paginationButton} accent>Next <IoCaretForward/></Button>)}
+
+
+							{props.persistenceState.value.tutorialIndex != undefined
+								&& props.persistenceState.value.tutorial
+								&& props.persistenceState.value.tutorialIndex == props.persistenceState.value.tutorial.length - 1
+								&& (<Button class={styles.paginationButton} onClick={() => {
+									props.showingTutorialWarning!.value = true
+								}} accent>Exit Tutorial</Button>)}
```
Comment (kognise) [inline src/components/popups-etc/help.tsx:106]: Thank you so much for making this work always!

File: src/components/popups-etc/help.tsx [hunk 10]
```diff
@@ -44,11 +78,35 @@ export default function Help(props: HelpProps) {
...
+							&& props.persistenceState.value.tutorial
+							&& props.persistenceState.value.tutorialIndex < props.persistenceState.value.tutorial.length - 1
+							&& (<Button onClick={nextPage} class={styles.paginationButton} accent>Next <IoCaretForward/></Button>)}
+
+
+							{props.persistenceState.value.tutorialIndex != undefined
+								&& props.persistenceState.value.tutorial
+								&& props.persistenceState.value.tutorialIndex == props.persistenceState.value.tutorial.length - 1
+								&& (<Button class={styles.paginationButton} onClick={() => {
+									props.showingTutorialWarning!.value = true
```
Comment (kognise) [inline src/components/popups-etc/help.tsx:105]:
```suggestion
									props.showingTutorialWarning?.value = true
```

the assertion might be code smell

File: src/pages/gallery/[filename].astro [hunk 11]
```diff
@@ -16,10 +16,16 @@ const res = await fetch(`https://raw.githubusercontent.com/hackclub/sprig/main/g
 if (!res.ok) return Astro.redirect('/404', 302)
 const code = await res.text()
 
+const fileRegexp = /.*\/(.+)-(\d+)\.md/;
```
Comment (kognise) [inline src/pages/gallery/[filename].astro:19]:
```suggestion
const fileRegexp = /^.*\/(.+)-(\d+)\.md$/
```

File: src/pages/gallery/[filename].astro [hunk 12]
```diff
@@ -16,10 +16,16 @@ const res = await fetch(`https://raw.githubusercontent.com/hackclub/sprig/main/g
...
+
 const files = await Astro.glob('/games/*.md')
-const tutorial = files
-	.find(file => file.file.endsWith(`/${filename}.md`))
-	?.compiledContent()
+let tutorial: string[] | undefined = files
+	.filter(file => {
+		let regexedFile = file.file.match(fileRegexp)
```
Comment (kognise) [inline src/pages/gallery/[filename].astro:24]:
```suggestion
		const regexedFile = file.file.match(fileRegexp)
```

File: src/pages/~/[id].astro [hunk 13]
```diff
@@ -14,12 +14,19 @@ if (!session) return Astro.redirect(`/login?to=${Astro.request.url}`, 302)
 const game = await getGame(Astro.params.id!)
 if (!game || game.ownerId !== session.user.id) return Astro.redirect('/404', 302)
 
-let tutorial: string | undefined = undefined
+const fileRegexp = /.*\/(.+)-(\d+)\.md/;
```
Comment (kognise) [inline src/pages/~/[id].astro:17]:
```suggestion
const fileRegexp = /^.*\/(.+)-(\d+)\.md$/
```

etc.

File: src/components/popups-etc/help.module.css
Comment (kognise) [reply src/components/popups-etc/help.module.css:1]: Sorry, wrong file! I am referring to the tutorial warning popup thingy.

**Review by kognise — CHANGES_REQUESTED** [review]
Great job! Besides the comments I left, code is good. (or at least, as fine as all the other code. the spaghetti is my fault, not yours. mostly.)

do need to take a look at the styles in a deployed or local version before merging

File: src/pages/~/[id].astro [hunk 14]
```diff
@@ -14,12 +14,19 @@ if (!session) return Astro.redirect(`/login?to=${Astro.request.url}`, 302)
...
+let tutorial: string[] | undefined;
+let tutorialIndex: number | undefined;
 if (game.tutorialName) {
 	const files = await Astro.glob('/games/*.md')
-	tutorial = files
-		.find(file => file.file.endsWith(`/${game.tutorialName}.md`))
-		?.compiledContent()
+		
+		tutorial = files.filter(file => {
+			let regexedFile = file.file.match(fileRegexp)
```
Comment (kognise) [inline src/pages/~/[id].astro:25]:
```suggestion
			const regexedFile = file.file.match(fileRegexp)
```

File: games/maze_game_starter-0.md
Comment (kognise) [reply games/maze_game_starter-0.md:1]: not outdated

**Comment by recursiveforte** [comment]
> is there somewhere i can see a tutorial that has pagination so i can preview this?

https://sprig-git-cheru-temp-fork.hackclub.dev/gallery/maze_game_starter

File: src/components/popups-etc/help.tsx (hunk 9 above)
Comment (recursiveforte) [reply src/components/popups-etc/help.tsx:106]: np!
btw, optional chaining doesn't work on the left side of an assignment.

**Comment by recursiveforte** [comment]
is there a formatter that you used that I should use when developing?

**Comment by kognise** [comment]
> is there a formatter that you used that I should use when developing?

not in this project, although maybe it warrants one now that there are multiple devs. in general, when lacking a formatter i just recommend doing your best to replicate the style of the code around you

**Comment by kognise** [comment]
> > is there somewhere i can see a tutorial that has pagination so i can preview this?
> 
> [sprig-git-cheru-temp-fork.hackclub.dev/gallery/maze_game_starter](https://sprig-git-cheru-temp-fork.hackclub.dev/gallery/maze_game_starter)

thank you! i'm gonna repeat the feedback i gave irl, otherwise ready to merge:

- the pagination container should have `justify-content: center` or `space-between` (whatever you think looks better)
- please make the back button not have a background or be less visually obtrusive

**Pull request merged**